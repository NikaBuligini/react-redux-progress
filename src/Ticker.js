/* eslint-disable no-console */
// @flow

import React from 'react';

const MAX_PERCENT = 85;

const isDevelopment =
  ['production', 'test'].indexOf(process.env.NODE_ENV) === -1;

const warn = (...args) => {
  if (isDevelopment) {
    console.warn(...args);
  }
};

/* eslint-disable react/no-unused-prop-types */
type Props = {
  percent: number,
  updateProgress: (progress: number) => void,
  autoIncrement: boolean,
  intervalTime: number,
  renderProgress: (percent: number) => React$Element<*>,
  maxPercent: number,
};
/* eslint-enable react/no-unused-prop-types */

type State = {
  percent: number,
  maxPercent: number,
};

class Ticker extends React.Component<Props, State> {
  static defaultProps = {
    percent: -1,
    autoIncrement: true,
    intervalTime: 450,
    maxPercent: MAX_PERCENT,
    renderProgress: () => null,
  };

  state = this.getInitialState(this.props);

  getInitialState(props: Props): State {
    let { maxPercent } = props;

    if (
      typeof maxPercent !== 'number' ||
      (maxPercent < 1 || maxPercent > 100)
    ) {
      warn('fallback to default maxPercent');
      maxPercent = MAX_PERCENT;
    }

    return {
      maxPercent,
      percent: this.props.percent,
    };
  }

  componentDidMount() {
    this.handleProps(this.props);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.interval) {
      // stop progress when new props come in.
      clearInterval(this.interval);
      this.interval = undefined;
    }

    if (this.timeout) {
      // clear timeout when new props come in.
      clearTimeout(this.timeout);
      this.timeout = undefined;
    }

    // start progress with updated props.
    this.handleProps(nextProps);
  }

  componentWillUnmount() {
    // cleaning up interval and timeout.
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = undefined;
    }

    if (this.timeout) {
      clearTimeout(this.timeout);
      this.timeout = undefined;
    }
  }

  increment = () => {
    let { percent } = this.state;

    percent += (Math.random() + 1 - Math.random()) * 3; // eslint-disable-line
    percent = percent < this.state.maxPercent ? percent : this.state.maxPercent;

    this.setState({ percent });
  };

  interval: ?Interval;

  handleProps = (props: Props) => {
    const { autoIncrement, percent, intervalTime } = props;

    if (autoIncrement && percent >= 0 && percent < 99) {
      this.interval = setInterval(this.increment, intervalTime);
    }

    if (percent >= 100) {
      this.setState({ percent: 99.9 }, () => {
        this.timeout = setTimeout(() => {
          this.setState({ percent: -1 }, () => this.props.updateProgress(-1));
        }, 400);
      });
    } else {
      this.setState({ percent });
    }
  };

  timeout: ?Timeout;

  render() {
    return this.props.renderProgress(this.state.percent);
  }
}

export default Ticker;
