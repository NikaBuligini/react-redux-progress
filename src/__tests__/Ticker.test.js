import React from 'react';
import sinon from 'sinon';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Ticker from '../Ticker';

let clock = null;

describe('<Ticker />', () => {
  beforeEach(() => {
    clock = sinon.useFakeTimers();
  });

  afterEach(() => {
    clock.restore();
  });

  it('should initially render empty progress', () => {
    const component = renderer.create(<Ticker />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should initially render progress html', () => {
    const component = renderer.create(
      <Ticker renderProgress={percent => <div>Progress {percent}</div>} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call renderProgress function', () => {
    const renderProgress = jest.fn();
    renderProgress.mockReturnValue(null);
    mount(<Ticker renderProgress={renderProgress} />);
    expect(renderProgress).toHaveBeenCalled();
  });

  it('should set state.percent as props.percent', () => {
    const expected = 50;
    const renderedComponent = mount(<Ticker percent={expected} />);
    expect(renderedComponent.state().percent).toEqual(expected);
  });

  it('should set state.maxPercent as props.maxPercent', () => {
    const expected = 50;
    const renderedComponent = mount(<Ticker maxPercent={expected} />);
    expect(renderedComponent.state().maxPercent).toEqual(expected);
  });

  it('should fallback to default maxPercent', () => {
    const passed = 150;
    const expected = 85;
    const renderedComponent = mount(<Ticker maxPercent={passed} />);
    expect(renderedComponent.state().maxPercent).toEqual(expected);
  });

  it('should fallback to default maxPercent when string is passed', () => {
    const passed = '95';
    const expected = 85;
    const renderedComponent = mount(<Ticker maxPercent={passed} />);
    expect(renderedComponent.state().maxPercent).toEqual(expected);
  });

  it('should call componentDidMount', () => {
    sinon.spy(Ticker.prototype, 'componentDidMount');
    // eslint-disable-next-line
    const renderedComponent = mount(
      <Ticker percent={0} updateProgress={noop => noop} />,
    );
    expect(Ticker.prototype.componentDidMount.calledOnce).toEqual(true);
    Ticker.prototype.componentDidMount.restore();
  });

  it('should call componentWillReceiveProps', () => {
    sinon.spy(Ticker.prototype, 'componentWillReceiveProps');
    const renderedComponent = mount(
      <Ticker percent={0} updateProgress={noop => noop} />,
    );
    renderedComponent.setProps({ percent: 50 });
    expect(Ticker.prototype.componentWillReceiveProps.calledOnce).toEqual(true);
    Ticker.prototype.componentWillReceiveProps.restore();
  });

  it('should unset Ticker.interval after getting new props', () => {
    const renderedComponent = mount(
      <Ticker percent={0} updateProgress={noop => noop} />,
    );
    const inst = renderedComponent.instance();

    clock.tick(1000);
    expect(inst.interval).toBeDefined();
    inst.componentWillReceiveProps({ percent: 50 });
    expect(inst.interval).toBeUndefined();
  });

  it('should unset Ticker.timeout after getting new props', () => {
    const renderedComponent = mount(
      <Ticker percent={100} updateProgress={noop => noop} />,
    );
    const inst = renderedComponent.instance();

    clock.tick(1000);
    expect(inst.timeout).toBeDefined();
    inst.componentWillReceiveProps({ percent: 50 });
    expect(inst.timeout).toBeUndefined();
  });

  it('should set state to -1 after new route mounts', () => {
    const renderedComponent = mount(
      <Ticker percent={0} updateProgress={noop => noop} />,
    );
    renderedComponent.setProps({ percent: 100 });
    clock.tick(501);
    expect(renderedComponent.state().percent).toEqual(-1);
  });

  it('should call componentWillUnmount', () => {
    sinon.spy(Ticker.prototype, 'componentWillUnmount');
    const renderedComponent = mount(
      <Ticker percent={0} updateProgress={noop => noop} />,
    );
    renderedComponent.unmount();
    expect(Ticker.prototype.componentWillUnmount.calledOnce).toEqual(true);
    Ticker.prototype.componentWillUnmount.restore();
  });

  it('should unset Ticker.interval after unmounting', () => {
    sinon.spy(Ticker.prototype, 'componentWillUnmount');
    const renderedComponent = mount(
      <Ticker percent={0} updateProgress={noop => noop} />,
    );
    const inst = renderedComponent.instance();

    clock.tick(1000);
    expect(inst.interval).toBeDefined();
    renderedComponent.unmount();
    expect(inst.interval).toBeUndefined();
    Ticker.prototype.componentWillUnmount.restore();
  });

  it('should unset Ticker.timeout after unmounting', () => {
    sinon.spy(Ticker.prototype, 'componentWillUnmount');
    const renderedComponent = mount(
      <Ticker percent={100} updateProgress={noop => noop} />,
    );
    const inst = renderedComponent.instance();

    clock.tick(1000);
    expect(inst.timeout).toBeDefined();
    renderedComponent.unmount();
    expect(inst.timeout).toBeUndefined();
    Ticker.prototype.componentWillUnmount.restore();
  });

  describe('increment progress', () => {
    beforeEach(() => {
      clock = sinon.useFakeTimers();
    });

    afterEach(() => {
      clock.restore();
    });

    it('should start incrementing progress if 0 <= percent < 100', () => {
      const initialPercent = 50;
      const renderedComponent = mount(
        <Ticker percent={initialPercent} updateProgress={noop => noop} />,
      );
      clock.tick(1000);
      expect(renderedComponent.state().percent).toBeGreaterThan(initialPercent);
    });

    it('should stop incrementing progress if percent >= 100', () => {
      const initialPercent = 100;
      const expected = -1;
      const renderedComponent = mount(
        <Ticker percent={initialPercent} updateProgress={noop => noop} />,
      );
      clock.tick(1000);
      expect(renderedComponent.state().percent).toEqual(expected);
    });
  });
});
