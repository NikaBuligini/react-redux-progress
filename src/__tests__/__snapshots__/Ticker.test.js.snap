// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ticker /> should initially render empty progress 1`] = `null`;

exports[`<Ticker /> should initially render progress html 1`] = `
<div>
  Progress 
  -1
</div>
`;
